mithril-relay/src/p2p/peer.rs

```diff
@@ -0,0 +1,208 @@
+        }
+
+        let _listener_id = swarm.listen_on(self.addr.clone())?;
+        self.swarm = Some(swarm);
+
+        loop {
+            if let Some(PeerEvent::ListeningOnAddr { address }) = self.tick_swarm().await? {
+                info!("Peer: listening on"; "address" => format!("{address:?}"), "local_peer_id" => format!("{:?}", self.local_peer_id()));
+                self.addr_peer = Some(address);
+                break;
+            }
+        }
+
+        Ok(self)
+    }
+
+    /// Tick the peer swarm to receive the next event
+    pub async fn tick_swarm(&mut self) -> StdResult<Option<PeerEvent>> {
+        debug!("Peer: reading next event"; "local_peer_id" => format!("{:?}", self.local_peer_id()));
+        match self
+            .swarm
+            .as_mut()
+            .ok_or(PeerError::UnavailableSwarm())
+            .with_context(|| "Can not publish signature without swarm")?
+            .next()
+            .await
+        {
+            Some(swarm::SwarmEvent::NewListenAddr { address, .. }) => {
+                debug!("Peer: received listening address event"; "address" => format!("{address:?}"), "local_peer_id" => format!("{:?}", self.local_peer_id()));
+                Ok(Some(PeerEvent::ListeningOnAddr { address }))
+            }
+            Some(swarm::SwarmEvent::OutgoingConnectionError { peer_id, error, .. }) => {
+                debug!("Peer: received outgoing connection error event"; "error" => format!("{error:#?}"), "remote_peer_id" => format!("{peer_id:?}"), "local_peer_id" => format!("{:?}", self.local_peer_id()));
+                Ok(Some(PeerEvent::OutgoingConnectionError { peer_id, error }))
+            }
+            Some(swarm::SwarmEvent::ConnectionEstablished { peer_id, .. }) => {
+                debug!("Peer: received connection established event"; "remote_peer_id" => format!("{peer_id:?}"), "local_peer_id" => format!("{:?}", self.local_peer_id()));
+                Ok(Some(PeerEvent::ConnectionEstablished { peer_id }))
+            }
+            Some(swarm::SwarmEvent::Behaviour(event)) => {
+                debug!("Peer: received behaviour event"; "event" => format!("{event:#?}"), "local_peer_id" => format!("{:?}", self.local_peer_id()));
+                Ok(Some(PeerEvent::Behaviour { event }))
+            }
+            Some(event) => {
+                debug!("Peer: received other event"; "event" => format!("{event:#?}"), "local_peer_id" => format!("{:?}", self.local_peer_id()));
+                Ok(None)
+            }
+            _ => Ok(None),
+        }
+    }
+
+    /// Publish a signature on the P2P pubsub
+    pub fn publish_signature(
+        &mut self,
+        message: &RegisterSignatureMessage,
+    ) -> StdResult<gossipsub::MessageId> {
+        let topic = self
+            .topics
+            .get(MITHRIL_SIGNATURES_TOPIC_NAME)
+            .ok_or(PeerError::MissingTopic())
+            .with_context(|| "Can not publish signature on invalid topic")?
```

Comment:
Is it possible to specify the topic name in the error message?

Reply:
Done :heavy_check_mark: